src/browser/PasskeyUtils.cpp

```diff
@@ -0,0 +1,344 @@
+        && authenticatorAttachment != BrowserPasskeys::ATTACHMENT_CROSS_PLATFORM) {
+        return false;
+    }
+
+    const auto requireResidentKey = authenticatorSelection["requireResidentKey"].toBool();
+    if (requireResidentKey && !BrowserPasskeys::SUPPORT_RESIDENT_KEYS) {
+        return false;
+    }
+
+    const auto residentKey = authenticatorSelection["residentKey"].toString();
+    if (residentKey == "required" && !BrowserPasskeys::SUPPORT_RESIDENT_KEYS) {
+        return false;
+    }
+
+    if (residentKey.isEmpty() && requireResidentKey && !BrowserPasskeys::SUPPORT_RESIDENT_KEYS) {
+        return false;
+    }
+
+    const auto userVerification = authenticatorSelection["userVerification"].toBool();
+    if (userVerification && !BrowserPasskeys::SUPPORT_USER_VERIFICATION) {
+        return false;
+    }
+
+    return true;
+}
+
+bool PasskeyUtils::isRegistrableDomainSuffix(const QString& hostSuffixString, const QString& originalHost) const
+{
+    if (hostSuffixString.isEmpty()) {
+        return false;
+    }
+
+    if (!isDomain(originalHost)) {
+        return false;
+    }
+
+    const auto hostSuffix = QUrl::fromUserInput(hostSuffixString).host();
+    if (hostSuffix == originalHost) {
+        return true;
+    }
+
+    if (!isDomain(hostSuffix)) {
+        return false;
+    }
+
+    const auto prefixedHostSuffix = QString(".%1").arg(hostSuffix);
+    if (!originalHost.endsWith(prefixedHostSuffix)) {
+        return false;
+    }
+
+    if (hostSuffix == urlTools()->getTopLevelDomainFromUrl(hostSuffix)) {
+        return false;
+    }
+
+    const auto originalPublicSuffix = urlTools()->getTopLevelDomainFromUrl(originalHost);
+    if (originalPublicSuffix.isEmpty()) {
+        return false;
+    }
+
+    if (originalPublicSuffix.endsWith(prefixedHostSuffix)) {
+        return false;
+    }
+
+    if (!hostSuffix.endsWith(QString(".%1").arg(originalPublicSuffix))) {
+        return false;
+    }
+
+    return true;
+}
+
+bool PasskeyUtils::isDomain(const QString& hostName) const
+{
+    const auto domain = QUrl::fromUserInput(hostName).host();
+    return !domain.isEmpty() && !domain.endsWith('.') && Tools::isAsciiString(domain)
+           && !urlTools()->domainHasIllegalCharacters(domain) && !urlTools()->isIpAddress(hostName);
+}
+
+bool PasskeyUtils::isUserVerificationValid(const QString& userVerification) const
+{
+    return QStringList({BrowserPasskeys::REQUIREMENT_PREFERRED,
+                        BrowserPasskeys::REQUIREMENT_REQUIRED,
+                        BrowserPasskeys::REQUIREMENT_DISCOURAGED})
+        .contains(userVerification);
+}
+
+bool PasskeyUtils::isOriginAllowedWithLocalhost(bool allowLocalhostWithPasskeys, const QString& origin) const
+{
+    return (!allowLocalhostWithPasskeys && origin.startsWith("https://"))
+           || (allowLocalhostWithPasskeys && (origin.startsWith("https://") || origin.startsWith("http://localhost")));
```

Comment:
This would allow an origin like `http://localhosting.com`, which would allow MITM to occur when a user isn't expecting it. How about logic like:

```c++
if (origin.startsWith("https://") || (allowLocalhostWithPasskeys && origin.startsWith("file://")) {
  return true;
}

if (!allowLocalhostWithPasskeys) {
  return false;
}

const auto host = QUrl::fromUserInput(origin).host();

return host == "localhost" 
  || host == "localhost."
  || host.endsWith(".localhost")
  || host.endsWith(".localhost.");
```

Reply:
This could work.